shape guards

avikchaudhuri opened the pull request:
Summary: This PR introduces shape guards to export. Previously only value ranges,  equalities, and specializations would be tracked for symbolic expressions, and we had a forward hook to check them. Instead now we create a function to check shape guards and call it in the exported program.

Test Plan:
updated several tests

Rollback Plan:

Differential Revision: D80713603


cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @Lucaskabela

tugsbayasgalan commented on torch/export/_unlift.py:
````diff
@@ -73,20 +81,106 @@ def _check_inputs_match(args, kwargs, in_spec: pytree.TreeSpec) -> list:
+    ```
+    torch._assert(
+        args[2].size()[0] == 3,
+        "Guard failed: z['k'].size()[0] == 3",
````
I am surprised how it didn't trigger some retracing tests which throw RuntimeError's. (test_retrace_pre_autograd)

avikchaudhuri replied:
It did, I turned them into the corresponding "Guard failed" `AssertionError`s.

tugsbayasgalan commented on torch/export/_unlift.py:
```diff
@@ -73,20 +81,106 @@ def _check_inputs_match(args, kwargs, in_spec: pytree.TreeSpec) -> list:
+        a, s = c, c
+        for idx, path in enumerate(paths_of_placeholders):
+            # e.g., replace L['z']['k'] with args[2] for Python code (actual)
+            a = a.replace("L" + pytree.keystr(path), f"args[{idx}]")
```
I think this is what i will do probably for new export lol.

tugsbayasgalan commented on torch/export/_unlift.py:
```diff
@@ -73,20 +81,106 @@ def _check_inputs_match(args, kwargs, in_spec: pytree.TreeSpec) -> list:
+    # be marked impure to avoid being not cleaned up by DCE
+    guards_fn = GuardsFn()
+    guards_fn.forward = torch._dynamo.dont_skip_tracing(namespace["_"])  # type: ignore[call-overload, method-assign]
+    guards_fn._is_impure = True  # type: ignore[assignment]
```
I thought this was a thing only for fx.Nodes. Does it work on generic functions?

avikchaudhuri replied:
I believe `call_module` will look up `_is_impure` in the target. See https://github.com/pytorch/pytorch/blob/adae7f66aacf3f248c3101b858cf98d5809119fa/torch/fx/node.py#L810

tugsbayasgalan replied:
Ah i see.

pianpwk commented on torch/_export/utils.py:
```diff
@@ -419,7 +419,7 @@ def _check_symint(
         # we'll add a runtime assert checking equality to this replacement expression
         pass
-    elif arg != symint:
+    elif arg != int(symint):
```
hmm why is this needed now?

avikchaudhuri replied:
I forgot which test it was but there was a Proxy error I think without it

tugsbayasgalan commented on torch/export/_unlift.py:
```diff
@@ -489,4 +703,37 @@ def _unlift_exported_program_lifted_states(ep: ExportedProgram) -> torch.fx.Grap
+            graph
+        )
+        with graph.inserting_after(placeholders[-1]):
+            node = graph.call_module("_guards_fn", tuple(placeholders))
```
I think you need to mark `node` as is_impure here.

tugsbayasgalan commented on torch/export/_unlift.py:
````diff
@@ -73,20 +81,106 @@ def _check_inputs_match(args, kwargs, in_spec: pytree.TreeSpec) -> list:
+    )
+    ```
+
+    FAQ: Why do we generate code based on (flattened) args instead of
````
Maybe i should indeed just pass in shape env with flat_args directly here? So you dont need to manually flatten inputs again?

avikchaudhuri replied:
I need to flatten only to get the keys (i.e., paths to placeholders). If the keys were made available some other way maybe it would work too.

yushangdi commented on torch/_inductor/__init__.py:
```diff
@@ -292,6 +292,15 @@ def aot_compile(
+        gm.graph.erase_node(node)
+        delattr(gm, "_guards_fn")
+        gm.recompile()
+
```
regarding how AOTI handles input constraints:

- for unbacked symints, inductor re-generates them, so doesn't matter what export guards says
- for backed symints, the behavior is currently different in OSS and internal (because a flag is on by default in OSS and not on by default internally). If the flag is not on, we ignore them. If the flag is on, we look at each assert node in the fx graph, and convert that into assertions in cpp code. So if there's not assertion fx nodes in the graph, AOTI ignores them.

avikchaudhuri committed:
shape guards (#161178)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/161178

This PR introduces shape guards to export. Previously only value ranges,  equalities, and specializations would be tracked for symbolic expressions, and we had a forward hook to check them. Instead now we create a function to check shape guards and call it in the exported program.

Test Plan:
updated several tests

Rollback Plan:

Reviewed By: tugsbayasgalan

Differential Revision: D80713603
